File: bdp-devops/ambari/ambari-web/test/views/main/service/info/config_test.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


var App = require('app');
require('views/main/service/info/configs');
var batchUtils = require('utils/batch_scheduled_requests');

describe('App.MainServiceInfoConfigsView', function () {

  var view;

  beforeEach(function () {
    view = App.MainServiceInfoConfigsView.create({
      controller: Em.Object.create({
        loadStep: Em.K,
        clearStep: Em.K,
        content: Em.Object.create()
      })
    });
  });

  describe("#resetConfigTabSelection()", function () {
    var tab = Em.Object.create({
      serviceName: 'S1',
      isActive: true
    });

    beforeEach(function () {
      sinon.stub(App.Tab, 'find').returns([tab]);
      sinon.stub(view, 'updateComponentInformation');
    });

    afterEach(function () {
      App.Tab.find.restore();
      view.updateComponentInformation.restore();
    });

    it("isActive should be false", function () {
      view.set('controller.content', Em.Object.create({serviceName: 'S1'}));
      view.resetConfigTabSelection();
      expect(tab.get('isActive')).to.be.false;
    });
  });

  describe('#updateComponentInformation', function () {

    var testCases = [
      {
        title: 'if components absent then counters should be 0',
        content: {
          restartRequiredHostsAndComponents: {}
        },
        result: {
          componentsCount: 0,
          hostsCount: 0
        }
      },
      {
        title: 'if host doesn\'t have components then hostsCount should be 1 and componentsCount should be 0',
        content: {
          restartRequiredHostsAndComponents: {
            host1: []
          }
        },
        result: {
          componentsCount: 0,
          hostsCount: 1
        }
      },
      {
        title: 'if host has 1 component then hostsCount should be 1 and componentsCount should be 1',
        content: {
          restartRequiredHostsAndComponents: {
            host1: [{}]
          }
        },
        result: {
          componentsCount: 1,
          hostsCount: 1
        }
      }
    ];
    testCases.forEach(function (test) {
      it(test.title, function () {
        view.set('controller.content', test.content);
        view.updateComponentInformation();
        expect(view.get('componentsCount')).to.equal(test.result.componentsCount);
        expect(view.get('hostsCount')).to.equal(test.result.hostsCount);
      });
    });
  });

  describe("#didInsertElement()", function() {
    var mock = {
      isLoading: function () {
        return {
          done: function (callback) {
            callback();
          }
        }
      }
    };

    beforeEach(function() {
      sinon.stub(App.router, 'get').returns(mock);
      sinon.stub(view.get('controller'), 'loadStep');
      sinon.stub(view, 'resetConfigTabSelection');
      view.didInsertElement();
    });
    afterEach(function() {
      App.router.get.restore();
      view.get('controller').loadStep.restore();
      view.resetConfigTabSelection.restore();
    });

    it("loadStep should be called", function() {
      expect(view.get('controller').loadStep.calledOnce).to.be.true;
    });

    it("resetConfigTabSelection should be called", function() {
      expect(view.resetConfigTabSelection.calledOnce).to.be.true;
    });
  });

  describe("#willDestroyElement()", function() {

    beforeEach(function() {
      sinon.stub(view.get('controller'), 'clearStep');
    });
    afterEach(function() {
      view.get('controller').clearStep.restore();
    });

    it("resetConfigTabSelection should be called", function() {
      view.willDestroyElement();
      expect(view.get('controller').clearStep.calledOnce).to.be.true;
    });
  });

  describe("#rollingRestartSlaveComponentName", function() {

    beforeEach(function() {
      sinon.stub(batchUtils, 'getRollingRestartComponentName', function(input) {
        return input;
      });
    });
    afterEach(function() {
      batchUtils.getRollingRestartComponentName.restore();
    });

    it("should return service name", function() {
      view.set('controller.content.serviceName', 'S1');
      view.propertyDidChange('rollingRestartSlaveComponentName');
      expect(view.get('rollingRestartSlaveComponentName')).to.equal('S1');
    });
  });

  describe("#rollingRestartActionName", function() {

    beforeEach(function() {
      sinon.stub(App.format, 'role', function(input) {
        return input;
      });
    });
    afterEach(function() {
      App.format.role.restore();
    });

    it("should return action name", function() {
      view.reopen({
        rollingRestartSlaveComponentName: 'C1'
      });
      view.propertyDidChange('rollingRestartActionName');
      expect(view.get('rollingRestartActionName')).to.equal(Em.I18n.t('rollingrestart.dialog.title').format('C1'));
    });

    it("should return empty", function() {
      view.reopen({
        rollingRestartSlaveComponentName: null
      });
      view.propertyDidChange('rollingRestartActionName');
      expect(view.get('rollingRestartActionName')).to.be.empty;
    });
  });
});

File: dataOds/InLong-versions/apache-inlong-1.2.0-incubating/inlong-tubemq/tubemq-client-twins/tubemq-client-go/config/config_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package config

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestParseAddress(t *testing.T) {
	address := "127.0.0.1:9092,127.0.0.1:9093?topic=Topic1&filters=12312323&filters=1212&topic=Topic2&filters=121212&filters=2321323&group=Group&tlsEnable=false&msgNotFoundWait=10000&heartbeatMaxRetryTimes=6"
	topicFilters := make(map[string][]string)
	topicFilters["Topic1"] = []string{"12312323", "1212"}
	topicFilters["Topic2"] = []string{"121212", "2321323"}
	c, err := ParseAddress(address)
	assert.Nil(t, err)
	assert.Equal(t, c.Consumer.Masters, "127.0.0.1:9092,127.0.0.1:9093")
	assert.Equal(t, c.Consumer.Topics, []string{"Topic1", "Topic2"})
	assert.Equal(t, c.Consumer.TopicFilters, topicFilters)
	assert.Equal(t, c.Consumer.Group, "Group")
	assert.Equal(t, c.Consumer.MsgNotFoundWait, 10000*time.Millisecond)

	assert.Equal(t, c.Net.TLS.Enable, false)

	assert.Equal(t, c.Heartbeat.MaxRetryTimes, 6)

	address = ""
	_, err = ParseAddress(address)
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topics=Topic&ttt"
	_, err = ParseAddress(address)
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topics=Topic&ttt=ttt"
	_, err = ParseAddress(address)
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic1&filters=12312323&filters=1212"
	c, err = ParseAddress(address)
	delete(topicFilters, "Topic2")
	assert.Nil(t, err)
	assert.Equal(t, c.Consumer.Topics, []string{"Topic1"})
	assert.Equal(t, c.Consumer.TopicFilters, topicFilters)

	address = "127.0.0.1:9092,127.0.0.1:9093?filters=1231232"
	_, err = ParseAddress(address)
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?filters=12312323&filters=1212"
	_, err = ParseAddress(address)
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=-1"
	c, err = ParseAddress(address)
	assert.Nil(t, err)
	err = c.ValidateConsumer()
	assert.Nil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=1"
	c, err = ParseAddress(address)
	assert.Nil(t, err)
	err = c.ValidateConsumer()
	assert.Nil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=0"
	c, err = ParseAddress(address)
	assert.Nil(t, err)
	err = c.ValidateConsumer()
	assert.Nil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=-2"
	c, err = ParseAddress(address)
	assert.Nil(t, err)
	err = c.ValidateConsumer()
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=2"
	c, err = ParseAddress(address)
	assert.Nil(t, err)
	err = c.ValidateConsumer()
	assert.NotNil(t, err)

	address = "127.0.0.1:9092,127.0.0.1:9093?topic=Topic&group=group&filters=12312323&consumePosition=a"
	c, err = ParseAddress(address)
	assert.NotNil(t, err)
}

func TestValidateGroup(t *testing.T) {
	config := NewDefaultConfig()
	err := config.validateGroup("")
	assert.NotNil(t, err)
	err = config.validateGroup("123-456[]")
	assert.NotNil(t, err)
	err = config.validateGroup("test123_456")
	assert.Nil(t, err)
	group := ""
	for i := 0; i < 1025; i++ {
		group += "t"
	}
	err = config.validateGroup(group)
	assert.NotNil(t, group)
}

func TestValidateConsumer(t *testing.T) {
	c := NewDefaultConfig()

	assert.NotNil(t, c.ValidateConsumer())
	WithConsumerMasters("masters")(c)
	assert.NotNil(t, c.ValidateConsumer())

	WithConsumerMasters("127.0.0.1:8000")(c)
	WithGroup("test-group")(c)
	assert.NotNil(t, c.ValidateConsumer())

	WithGroup("test_group")(c)
	assert.Nil(t, c.ValidateConsumer())

	WithTopics([]string{"topic1", "topic2"})(c)
	assert.Nil(t, c.ValidateConsumer())

	WithTopics([]string{"1topic, topic2"})(c)
	assert.NotNil(t, c.ValidateConsumer())

	topicFilters := map[string][]string{"1topic": {"filter1", "filter2"}, "topic2": {"filter3", "filter4"}}
	WithTopicFilters(topicFilters)(c)
	assert.NotNil(t, c.ValidateConsumer())

	topicFilters = map[string][]string{"topic1": {"-=-", "filter2"}, "topic2": {"filter3", "filter4"}}
	WithTopicFilters(topicFilters)(c)
	assert.NotNil(t, c.ValidateConsumer())

	topicFilters = map[string][]string{"topic1": {"filter1", "filter2"}, "topic2": {"filter3", "filter4"}}
	WithTopicFilters(topicFilters)(c)
	assert.Nil(t, c.ValidateConsumer())

	partitionOffset := map[string]int64{"181895251:test_1": 0, "181895251:test_2": 10}
	WithBoundConsume("", 0, true, partitionOffset)(c)
	assert.NotNil(t, c.ValidateConsumer())

	WithBoundConsume("11", 0, true, partitionOffset)(c)
	assert.NotNil(t, c.ValidateConsumer())

	partitionOffset = map[string]int64{"181895251:test_1:1": 0, "181895251:test_2:2": 10}
	WithBoundConsume("11", 0, true, partitionOffset)(c)
	assert.NotNil(t, c.ValidateConsumer())

	partitionOffset = map[string]int64{"181895251:topic1:1": 0, "181895251:topic2:2": 10}
	WithBoundConsume("11", 0, true, partitionOffset)(c)
	assert.Nil(t, c.ValidateConsumer())

	WithConsumePosition(2)(c)
	assert.NotNil(t, c.ValidateConsumer())

	WithConsumePosition(-2)(c)
	assert.NotNil(t, c.ValidateConsumer())

	WithConsumePosition(-1)(c)
	assert.Nil(t, c.ValidateConsumer())

	WithConsumePosition(0)(c)
	assert.Nil(t, c.ValidateConsumer())

	WithConsumePosition(1)(c)
	assert.Nil(t, c.ValidateConsumer())
}
